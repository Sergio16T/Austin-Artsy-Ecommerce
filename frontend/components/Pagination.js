import React, { Component } from 'react';
import PaginationStyles from './styles/PaginationStyles'; 
import { Query } from 'react-apollo'; 
import Head from "next/head"; 
import Link from "next/link"; 
import styled from 'styled-components'; 
import gql from 'graphql-tag'; 

const perPage = 4; 

const PAGINATION_QUERY = gql`
    query PAGINATION_QUERY{
        itemsConnection {
            aggregate {
                count
            }
        }
    }
`;


const PaginationContainer = styled.div`
    padding-top: 4rem; 
    width: 100%; 
    display: flex; 
    justify-content: center; 
`;
const Placeholder = styled.div`
    height: 45px;
    margin-top: 24px; 
    margin-bottom: 24px; 
`; 
const Pagination = props => {
    return (
        <PaginationContainer>
            <Query query={PAGINATION_QUERY}>
                {({data, error, loading}) => {
                    if(loading) return <Placeholder/>; 
                    if(error) return <p>{error.message}</p>
                    const count = data.itemsConnection.aggregate.count; 
                    const pages = Math.ceil(count / perPage); 
                    return (
                    <PaginationStyles data-testid="pagination">
                        <Head>
                            <title>Austin's List page {props.page} of {pages}</title>
                        </Head>
                        <Link                      
                        href={{
                            pathname: "/items",
                            query: {page: props.page -1}
                        }}
                        >
                        <a aria-disabled={props.page <= 1} id="prev"> &#8592; Prev</a>
                        </Link>
                        <p data-test="pagecount">Page {props.page} of {pages}</p>
                        <Link                      
                        href={{
                            pathname:"/items",
                            query: {page: props.page + 1}
                        }}

                        >
                        <a aria-disabled={props.page >= pages} id="next">Next &#8594;</a>
                        </Link>
                    </PaginationStyles>
                    )}}
            </Query> 
        </PaginationContainer>
    );
};


export default Pagination;
export { PAGINATION_QUERY }; 