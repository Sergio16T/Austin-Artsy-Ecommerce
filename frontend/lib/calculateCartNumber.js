export default function(cart) {
    return cart.reduce((startValue, element)=> {
        return startValue + element.quantity; 
    }, 0); 
}