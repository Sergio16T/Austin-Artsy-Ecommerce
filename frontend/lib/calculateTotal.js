export default function(cart) {
    return cart.reduce((startValue, element) => {
    if(!element.item) return startValue; 
    return startValue + (element.quantity * element.item.price)}, 0); 
    
}