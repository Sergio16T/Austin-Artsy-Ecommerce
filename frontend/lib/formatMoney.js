export default function(amount) {
    let price = amount / 100; 
    return new Intl.NumberFormat('en', {
        style: "currency",
        currency: "USD",
    }).format(price);
}

/* when receiving money from form make sure it always has two floating point numbers and then just multiply it by 100
to store it in graphql as cents in int type then when I'm displaying money multiply it by 100.
that way I can format it for use later.. so it's always going to be handled in cents 
165.33 * 100
16533
165 * 100
16500
16500/100
165
16533/100
165.33
 */ 
