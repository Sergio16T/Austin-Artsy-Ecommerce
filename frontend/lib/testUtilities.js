import casual from 'casual'; 
//.seed() is to provide ability to use a specific seed in order to get a repeatable random sequence for our testing purposes
casual.seed(111); 

const mockItem = () => ({
    // __typename: 'Item',
    id: '123',
    price: 5000,
    user: null,
    image: 'dog.jpg',
    title: 'dogs are the best',
    description: 'dogs description',
    largeImage: 'dog-large.jpg',
}); 

const mockUserWithItemsInCart = () =>  ({
    __typename: 'User', 
    id: "123", 
    name: casual.name, 
    email: casual.email, 
    permissions: ["ADMIN"], 
    cart: [{
        id: 'abc', 
        quantity: 2, 
        item: {
            id: 'dgh', 
            price: 15000, 
            image: 'www.dog.image.com',
            title: "puppy"
        }
        },
    ]
}); 
const mockUser = () => ({
    __typename: 'User', 
    id: "123", 
    name: casual.name, 
    email: casual.email, 
    permissions: ['ADMIN'], 
    cart: [],
    orders: []
}); 
const mockOrderItem = () => ({
    __typename: 'OrderItem',
    id: casual.uuid,
    image: `${casual.word}.jpg`,
    title: casual.words(),
    price: 4234,
    quantity: 1,
    description: casual.words(),
  });
  
  const mockOrder = () => ({
    __typename: 'Order',
    id: 'ord123',
    charge: 'ch_123',
    total: 40000,
    items: [mockOrderItem(), mockOrderItem()],
    createdAt: '2018-04 - 06T19: 24: 16.000Z',
    user: mockUser(),
  });
  
  const mockCartItem = overrides => ({
    __typename: 'CartItem',
    id: 'omg123',
    quantity: 3,
    item: mockItem(),
    user: mockUser(),
    ...overrides,
  });
//mock LocalStorage 
class MockLocalStorage {
    constructor(){
        this.store={}; 
    }

    clear() {
        this.store = {}; 
    }
    getItem(key) {
        return this.store[key] || null; 
    }
    setItem(key, value) {
        this.store[key] = value.toString(); 
    }
    removeItem(key) {
        delete this.store[key]; 
    }
}

export { 
    mockItem, 
    MockLocalStorage, 
    mockUser, 
    mockCartItem, 
    mockOrderItem, 
    mockOrder }; 