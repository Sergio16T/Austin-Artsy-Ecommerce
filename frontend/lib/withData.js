import withApollo from 'next-with-apollo'; 
import ApolloClient from 'apollo-boost'; 
import { endpoint } from '../config'; 

function createClient({ headers}) {
    return new ApolloClient({
        uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint, 
        // on each request pass the headers and setContext change credentials to 'include' for whether the user agent should send cookies from other domains in case of CORS
        request: operation => {
            operation.setContext({
                fetchOptions: {
                    credentials: 'include', // logged in cookies present ? include.  (default value is same-origin) 
                },
                headers, 
            }); 
        },
    }); 
}

export default withApollo(createClient); 