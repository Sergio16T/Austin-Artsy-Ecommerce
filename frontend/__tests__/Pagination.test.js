import { render, screen, act } from '@testing-library/react'; 
import Pagination, { PAGINATION_QUERY } from '../components/Pagination'; 
import { MockedProvider } from '@apollo/react-testing'; 
import { mockUser, mockCartItem } from '../lib/testUtilities'; 
import wait from 'waait'; 

const mockPagination = (length) => {
    return [{
        request: { query: PAGINATION_QUERY }, 
        result: { 
            data: {
                itemsConnection: {
                    __typename: 'aggregate',
                    aggregate: {
                        __typename: 'count', 
                        count: length
                    }
                },
            },
        },
    }]
}

describe('<Pagination/>', () => {
    it('mounts with correct number of pages', async () => {
        const { container } = render(
            <MockedProvider mocks={mockPagination(18)}>
                <Pagination page={1}/>
            </MockedProvider>
        ); 
        await act( async() => {
            await wait(); 
        }); 
        expect(screen.getByText("Page 1 of 5")).toBeInTheDocument(); 
        expect(container).toMatchSnapshot(); 
    }); 
    it('disables prev button when on page 1', async () => {
        const { container, findByTestId, debug, getByText } = render(
            <MockedProvider mocks={mockPagination(10)}>
                <Pagination page={1}/>
            </MockedProvider>
        ); 
        await findByTestId("pagination"); 
        // debug(); 
        expect(getByText(/Prev/)).toHaveAttribute('aria-disabled', 'true'); 
        expect(getByText(/Next/)).toHaveAttribute('aria-disabled', 'false'); 
    });
    it('disables next button when on last page', async () => {
        const { container, findByTestId, debug, getByText } = render(
            <MockedProvider mocks={mockPagination(10)}>
                <Pagination page={3}/>
            </MockedProvider>
        ); 
        await findByTestId("pagination"); 
        // debug(); 
        expect(getByText(/Prev/)).toHaveAttribute('aria-disabled', 'false'); 
        expect(getByText(/Next/)).toHaveAttribute('aria-disabled', 'true'); 
    });

}); 