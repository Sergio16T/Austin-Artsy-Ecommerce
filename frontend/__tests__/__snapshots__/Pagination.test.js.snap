// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> mounts with correct number of pages 1`] = `
<div>
  <div
    class="Pagination__PaginationContainer-sc-19vuk7g-0 kTjDNS"
  >
    <div
      class="PaginationStyles-cr7hxd-0 hzzTYo"
      data-testid="pagination"
    >
      <a
        aria-disabled="true"
        href="/items?page=0"
        id="prev"
      >
         ← Prev
      </a>
      <p
        data-test="pagecount"
      >
        Page 
        1
         of 
        5
      </p>
      <a
        aria-disabled="false"
        href="/items?page=2"
        id="next"
      >
        Next →
      </a>
    </div>
  </div>
</div>
`;
